lib/descriptor/descriptor.js

```diff
@@ -0,0 +1,176 @@
+/*!
+ * descriptor.js - descriptor object for bcoin
+ * Copyright (c) 2023, the bcoin developers (MIT License).
+ * https://github.com/bcoin-org/bcoin
+ */
+
+'use strict';
+
+const {stringType, addChecksum} = require('./common');
+// const assert = require('bsert');
+
+/**
+ * Descriptor
+ * Base class for descriptors.
+ * Represents an output script
+ * @property {String} name
+ * @property {PubkeyProvider[]} pubkeys
+ * @property {Descriptor[]} subdescriptors
+ * @property {Network} network
+ */
+
+class Descriptor {
+  /**
+   * Create a descriptor
+   * @constructor
+   */
+
+  constructor() {
+    this.name = null;
+    this.pubkeys = [];
+    this.subdescriptors = [];
+    this.network = null;
+  }
+
+  /**
+   * Test whether the descriptor contains a private key.
+   * @returns {Boolean}
+   */
+
+  hasPrivateKeys() {
+    for (const key of this.pubkeys) {
+      if (key.getPrivatekey() !== null) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Test whether the descriptor contains public/private keys
+   * in the form of HD chains
+   * @returns {Boolean}
+   */
+
+  isRange() {
+    for (const key of this.pubkeys) {
+      if (key.isRange()) {
+        return true;
+      }
+    }
+    for (const subdesc of this.subdescriptors) {
+      if (subdesc.isRange()) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Whether this descriptor has all information about signing
+   * (igonoring private keys).
+   * Returns false only for `addr` and `raw` type.
+   * @returns {Boolean}
+   */
+
+  isSolvable() {
+    for (const subdesc of this.subdescriptors) {
+      if (!subdesc.isSolvable()) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  toStringExtra() {
```

Comment:
Add documentation for un-implemented functions. Helps keep track of what needs to be done and why these functions should exist.